osu.Game.Rulesets.Osu/Scoring/OsuScoreProcessor.cs

```diff
@@ -17,14 +17,21 @@ public OsuScoreProcessor()
         }
 
         public override ScoreRank RankFromScore(double accuracy, IReadOnlyDictionary<HitResult, int> results)
-        {
-            ScoreRank rank = base.RankFromScore(accuracy, results);
+            => adjustRank(base.RankFromScore(accuracy, results), results.GetValueOrDefault(HitResult.Miss));
+
+        protected override ScoreRank MinimumRankFromScore(double accuracy, IReadOnlyDictionary<HitResult, int> results)
+            // this will be wrong when the remaining judgements do not affect accuracy and the player did not miss before,
+            // but it's a bit of a tiny detail to fix. we'll at least show an S/SS rank when the user completes the beatmap
+            // (especially when the beatmap has an outro storyboard).
+            => adjustRank(base.RankFromScore(accuracy, results), JudgedHits == MaxHits ? results.GetValueOrDefault(HitResult.Miss) : int.MaxValue);
```

Comment:
This comment says that "what" but not the "why". If I read this in isolation (which is the first thing I did when seeing the history of this PR), my first question is going to be "why will it be wrong?".

I also don't really follow why the special case logic can't remain in `RankFromScore` as it was before this PR. If i dig deeper I'm sure I can understand why, but it's not really passing the smell test at this point for sane code. Which is probably not great since these changes are being made in `ScoreProcessor`, not some isolated skinnable UI element class..

Reply:
This comment is written specifically to explain the "why" part: "why will it be wrong? because *when the remaining judgements do not affect accuracy and the player did not miss before* then the rank will be incorrectly A rather than S/SS".

This is `OsuScoreProcessor` so I would like to imagine there's a level of locality to this change. However, I can move this to become a workaround from a UI element point, sounds like a better plan to keep the `ScoreProcessor` class clean and consistent from changes like these.

Reply:
After further discussion, I understand where the concern comes from. The comment is meant to highlight the `JudgedHits == MaxHits` particularly, but it looks like it's trying to explain why this entire override exists. The comment is also written without setting up the scenario exactly, beginning from "it's wrong" but not...explaining why it's wrong.

Take a look at https://github.com/ppy/osu/pull/28614/commits/53272c16a0ca22a7dde9e7378c1199c9c1f6716c and see if it helps.